### ISSUE TITLE
shutting down threebot server with ctrl+c gives gedis KeyError

### ISSUE THREAD

**Issue by xmonader** [threefoldtech/js-sdk#276]
I started 3bot server and after a while I wanted graceful shutdown so in terminal i sent `ctrl+c`

```
^C^C2020-07-21 14:11:14.695 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
^C2020-07-21 14:11:16.372 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:11:17.714 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:11:18.857 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
2020-07-21 14:11:19.934 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
2020-07-21 14:11:19.943 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
Traceback (most recent call last):
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/gevent/hub.py", line 269, in handle
    self.handler(*self.args, **self.kwargs)
  File "/home/ahmed/wspace/js/js-sdk/jumpscale/servers/threebot/threebot.py", line 643, in stop
    self.rack.stop()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 78, in stop
    self._stop(server_name)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 21, in _stop
    self._started.remove(server_name)
KeyError: 'gedis'
2020-07-21T12:11:20Z
2020-07-21 14:11:21.139 | ERROR    | jumpscale.tools.errorhandler.errorhandler:_handle_exception:55 - KeyError: 'gedis'
Traceback (most recent call last):
> File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/bin/threebot", line 33, in <module>
    sys.exit(load_entry_point('js-sdk', 'console_scripts', 'threebot')())
    │   │    └ <function importlib_load_entry_point at 0x7fbea4527e18>
    │   └ <built-in function exit>
    └ <module 'sys' (built-in)>
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
           │    │     │       └ {}
           │    │     └ ()
           │    └ <function BaseCommand.main at 0x7fbe9eedd510>
           └ <Group cli>
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
         │    │      └ <click.core.Context object at 0x7fbea2e48128>
         │    └ <function MultiCommand.invoke at 0x7fbe9eede268>
         └ <Group cli>
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
           │               │       │       │      └ <click.core.Context object at 0x7fbea2220160>
           │               │       │       └ <function Command.invoke at 0x7fbe9eedde18>
           │               │       └ <Command start>
           │               └ <click.core.Context object at 0x7fbea2220160>
           └ <function MultiCommand.invoke.<locals>._process_result at 0x7fbea2dfbb70>
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           │   │      │    │           │   └ {'local': True, 'identity': None, 'background': False}
           │   │      │    │           └ <click.core.Context object at 0x7fbea2220160>
           │   │      │    └ <function start at 0x7fbe9ac53268>
           │   │      └ <Command start>
           │   └ <function Context.invoke at 0x7fbe9eedd048>
           └ <click.core.Context object at 0x7fbea2220160>
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
           │         │       └ {'local': True, 'identity': None, 'background': False}
           │         └ ()
           └ <function start at 0x7fbe9ac53268>

  File "/home/ahmed/wspace/js/js-sdk/jumpscale/entry_points/threebot.py", line 140, in start
    j.servers.threebot.start_default(wait=True, local=local)
    │ │                                         │     └ True
    │ │                                         └ True
    │ └ <property object at 0x7fbe9eee29f8>
    └ <jumpscale.loader.J object at 0x7fbe9ef189b0>

  File "/home/ahmed/wspace/js/js-sdk/jumpscale/servers/threebot/__init__.py", line 29, in start_default
    server.start(wait=wait)
    │      │     │    └ True
    │      │     └ True
    │      └ <function ThreebotServer.start at 0x7fbe989786a8>
    └ ThreebotServer(_Base__instance_name='default', _Base__parent=None, ___package_manager=<jumpscale.core.base.factory.jumpscales...

  File "/home/ahmed/wspace/js/js-sdk/jumpscale/servers/threebot/threebot.py", line 636, in start
    self.rack.start(wait=wait)  # to keep the server running
    │    │          │    └ True
    │    │          └ True
    │    └ <property object at 0x7fbe98974b88>
    └ ThreebotServer(_Base__instance_name='default', _Base__parent=None, ___package_manager=<jumpscale.core.base.factory.jumpscales...

  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 64, in start
    self._running_event.wait()
    │    │              └ <cyfunction Event.wait at 0x7fbe9ff53270>
    │    └ <gevent._event.Event object at 0x7fbe98aa9348>
    └ ServerRack(_Base__instance_name=None, _Base__parent=None, _factories={}, _running_event=<gevent._event.Event object at 0x7fbe...

  File "src/gevent/event.py", line 127, in gevent._event.Event.wait

  File "src/gevent/_abstract_linkable.py", line 192, in gevent.__abstract_linkable.AbstractLinkable._wait

  File "src/gevent/_abstract_linkable.py", line 165, in gevent.__abstract_linkable.AbstractLinkable._wait_core

  File "src/gevent/_abstract_linkable.py", line 169, in gevent.__abstract_linkable.AbstractLinkable._wait_core

  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch

  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch

  File "src/gevent/_greenlet_primitives.py", line 64, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch

  File "src/gevent/__greenlet_primitives.pxd", line 35, in gevent.__greenlet_primitives._greenlet_switch

KeyError: 'gedis'
Traceback (most recent call last):
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/gevent/hub.py", line 269, in handle
    self.handler(*self.args, **self.kwargs)
  File "/home/ahmed/wspace/js/js-sdk/jumpscale/servers/threebot/threebot.py", line 643, in stop
    self.rack.stop()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 78, in stop
    self._stop(server_name)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 21, in _stop
    self._started.remove(server_name)
KeyError: 'gedis'
2020-07-21T12:11:21Z
--- Logging error in Loguru Handler #1 ---
Record was: {'elapsed': datetime.timedelta(0, 513, 288827), 'exception': ExceptionRecattr(type=<class 'KeyError'>, value=KeyError('gedis',), traceback=<traceback object at 0x7fbe988f40c8>), 'extra': {'appname': 'threebot_default', 'category': '', 'data': None}, 'file': 'errorhandler.py', 'function': '_handle_exception', 'level': 'ERROR', 'line': 55, 'message': "KeyError: 'gedis'", 'module': 'errorhandler', 'name': 'jumpscale.tools.errorhandler.errorhandler', 'process': '30412', 'thread': '140456732462120', 'time': datetime(2020, 7, 21, 14, 11, 21, 139194, tzinfo=datetime.timezone(datetime.timedelta(0, 7200), 'EET'))}
Traceback (most recent call last):
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/loguru/_handler.py", line 144, in emit
    self._writer(str_record)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/core/logging/logging.py", line 166, in _handle
    if not self._db.is_running():
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/clients/redis/redis.py", line 51, in is_running
    return self.redis_client.ping()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/client.py", line 1378, in ping
    return self.execute_command('PING')
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/client.py", line 901, in execute_command
    return self.parse_response(conn, command_name, **options)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/client.py", line 915, in parse_response
    response = connection.read_response()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/connection.py", line 739, in read_response
    response = self._parser.read_response()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/connection.py", line 324, in read_response
    raw = self._buffer.readline()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/connection.py", line 256, in readline
    self._read_from_socket()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/connection.py", line 198, in _read_from_socket
    data = recv(self._sock, socket_read_size)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/redis/_compat.py", line 72, in recv
    return sock.recv(*args, **kwargs)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/gevent/_socket3.py", line 382, in recv
    self._wait(self._read_event)
  File "src/gevent/_hub_primitives.py", line 284, in gevent.__hub_primitives.wait_on_socket
  File "src/gevent/_hub_primitives.py", line 289, in gevent.__hub_primitives.wait_on_socket
  File "src/gevent/_hub_primitives.py", line 271, in gevent.__hub_primitives._primitive_wait
  File "src/gevent/_hub_primitives.py", line 46, in gevent.__hub_primitives.WaitOperationsGreenlet.wait
  File "src/gevent/_hub_primitives.py", line 46, in gevent.__hub_primitives.WaitOperationsGreenlet.wait
  File "src/gevent/_hub_primitives.py", line 55, in gevent.__hub_primitives.WaitOperationsGreenlet.wait
  File "src/gevent/_waiter.py", line 151, in gevent.__waiter.Waiter.get
  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 64, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/__greenlet_primitives.pxd", line 35, in gevent.__greenlet_primitives._greenlet_switch
KeyError: 'gedis'
--- End of logging error --
```

**Comment by xmonader** [comment]
interestingly happens only when sending multiple ctrl+c 

with single signal
```
  js-sdk--OGPzDz5-py3.6   …/js-sdk     development   threebot start --local
2020-07-21 14:18:09.799 | INFO     | jumpscale.entry_points.threebot:check_for_bins:30 - ✅ binaries nginx,redis-server,tmux,git required are installed.
2020-07-21 14:18:09.863 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2020-07-21 14:18:09.863 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: the configuration file /tmp/tmph1r4kehz syntax is ok
2020-07-21 14:18:09.863 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: configuration file /tmp/tmph1r4kehz test is successful
2020-07-21 14:18:14.525 | INFO     | jumpscale.core.application.application:start:23 - Application gedis is started, process id: 3044
2020-07-21 14:18:14.528 | INFO     | jumpscale.core.application.application:start:23 - Application threebot_default is started, process id: 3044
2020-07-21 14:18:14.530 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package auth
2020-07-21 14:18:15.613 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package chatflows
2020-07-21 14:18:15.842 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package admin
2020-07-21 14:18:16.034 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package weblibs
2020-07-21 14:18:16.070 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package tfgrid_solutions
2020-07-21 14:18:16.153 | INFO     | jumpscale.servers.threebot.threebot:_install_all:496 - Configuring package codeserver
2020-07-21 14:18:16.823 | INFO     | jumpscale.servers.threebot.threebot:start:630 - Reloading nginx
2020-07-21 14:18:16.868 | INFO     | jumpscale.servers.threebot.threebot:start:635 - Threebot is running at http://localhost:8080 and https://localhost:8443
^C2020-07-21 14:18:29.890 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:18:31.038 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
{'gedis', 'gedis_http', 'chatflows_main', 'auth_main'}
{'gedis_http', 'chatflows_main', 'auth_main'}
{'chatflows_main', 'auth_main'}
{'chatflows_main'}


```

multiple signals

```
  js-sdk--OGPzDz5-py3.6   …/js-sdk     development   threebot start --local
2020-07-21 14:18:38.350 | INFO     | jumpscale.entry_points.threebot:check_for_bins:30 - ✅ binaries nginx,redis-server,tmux,git required are installed.
2020-07-21 14:18:38.395 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: [alert] could not open error log file: open() "/var/log/nginx/error.log" failed (13: Permission denied)
2020-07-21 14:18:38.396 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: the configuration file /tmp/tmppgeo90__ syntax is ok
2020-07-21 14:18:38.396 | DEBUG    | jumpscale.entry_points.threebot:test_privileged_ports_bind:53 - nginx: configuration file /tmp/tmppgeo90__ test is successful
^C^C2020-07-21 14:18:40.149 | INFO     | jumpscale.core.application.application:stop:26 - Application init is stopped
2020-07-21 14:18:41.086 | INFO     | jumpscale.core.application.application:stop:26 - Application init is stopped
2020-07-21 14:18:44.189 | INFO     | jumpscale.core.application.application:start:23 - Application gedis is started, process id: 3478
2020-07-21 14:18:44.193 | INFO     | jumpscale.core.application.application:start:23 - Application threebot_default is started, process id: 3478
2020-07-21 14:18:44.195 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package auth
2020-07-21 14:18:47.103 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package chatflows
2020-07-21 14:18:47.331 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package admin
2020-07-21 14:18:47.526 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package weblibs
2020-07-21 14:18:47.558 | INFO     | jumpscale.servers.threebot.threebot:start:618 - Configuring package tfgrid_solutions
2020-07-21 14:18:47.632 | INFO     | jumpscale.servers.threebot.threebot:_install_all:496 - Configuring package codeserver
2020-07-21 14:18:48.266 | INFO     | jumpscale.servers.threebot.threebot:start:630 - Reloading nginx
2020-07-21 14:18:48.304 | INFO     | jumpscale.servers.threebot.threebot:start:635 - Threebot is running at http://localhost:8080 and https://localhost:8443
^C^C^C2020-07-21 14:18:51.469 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:18:51.471 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:18:52.588 | INFO     | jumpscale.core.application.application:stop:26 - Application threebot_default is stopped
2020-07-21 14:18:53.643 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
2020-07-21 14:18:54.730 | INFO     | jumpscale.servers.gedis.server:stop:209 - Shutting down...
{'chatflows_main', 'gedis', 'auth_main', 'gedis_http'}
{'chatflows_main', 'auth_main', 'gedis_http'}
{'chatflows_main', 'auth_main'}
{'chatflows_main'}
set()
Traceback (most recent call last):
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/gevent/hub.py", line 269, in handle
    self.handler(*self.args, **self.kwargs)
  File "/home/ahmed/wspace/js/js-sdk/jumpscale/servers/threebot/threebot.py", line 643, in stop
    self.rack.stop()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 79, in stop
    self._stop(server_name)
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/jumpscale/servers/rack/rack.py", line 22, in _stop
    self._started.remove(server_name)
KeyError: 'gedis'
2020-07-21T12:18:54Z
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/loguru/_logger.py", line 879, in remove
    handler.stop()
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/loguru/_handler.py", line 153, in stop
    with self._lock:
  File "src/gevent/_semaphore.py", line 141, in gevent.__semaphore.Semaphore.__enter__
  File "src/gevent/_semaphore.py", line 142, in gevent.__semaphore.Semaphore.__enter__
  File "src/gevent/_semaphore.py", line 100, in gevent.__semaphore.Semaphore.acquire
  File "/home/ahmed/.cache/pypoetry/virtualenvs/js-sdk--OGPzDz5-py3.6/lib/python3.6/site-packages/gevent/thread.py", line 84, in acquire
    return BoundedSemaphore.acquire(self, blocking, timeout)
  File "src/gevent/_semaphore.py", line 100, in gevent.__semaphore.Semaphore.acquire
  File "src/gevent/_semaphore.py", line 128, in gevent.__semaphore.Semaphore.acquire
  File "src/gevent/_abstract_linkable.py", line 192, in gevent.__abstract_linkable.AbstractLinkable._wait
  File "src/gevent/_abstract_linkable.py", line 165, in gevent.__abstract_linkable.AbstractLinkable._wait_core
  File "src/gevent/_abstract_linkable.py", line 169, in gevent.__abstract_linkable.AbstractLinkable._wait_core
  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 60, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/_greenlet_primitives.py", line 64, in gevent.__greenlet_primitives.SwitchOutGreenletWithLoop.switch
  File "src/gevent/__greenlet_primitives.pxd", line 35, in gevent.__greenlet_primitives._greenlet_switch
KeyError: 'gedis'

```

**Comment by xmonader** [comment]
fixed in 9ebaaaa0a0a667bb3e15e91d17c95916cb76293a

**Comment by Dina-Abd-Elrahman** [comment]
- verified.
